src/executor-group.jl

```diff
@@ -196,15 +216,161 @@ function set_params!(self::DataParallelExecutorGroup,
 end
 
 ##
-# Internals
+# Utility
 ##
 
+update_params(self::DataParallelExecutorGroup, updater, update_on_kvstore, kvstore::Void = nothing) = update_params(self, updater, update_on_kvstore, Nullable{KVStore}())
+update_params(self::DataParallelExecutorGroup, updater, update_on_kvstore, kvstore::KVStore) = update_params(self, updater, update_on_kvstore, Nullable(kvstore))
+function update_params(self::DataParallelExecutorGroup, updater, update_on_kvstore, kvstore::Nullable{KVStore})
+  num_dev = length(self.context)
+  for idx = 1:length(self.param_names)
+    #= if isa(self.grad_arrays[i][1], Void) =#
+    #=   continue =#
+    #= end =#
+    if in(idx, self.freeze_idx)
+      continue # Skip parameter update entirely
+    end
+    if !isnull(kvstore)
+      kvstore = get(kvstore)
+      # push gradient, priority is negative index
+      push!(kvstore, idx, self.param_arrays[idx], priority=-idx)
+      if update_on_kvstore
+        # pull back the weights
+        pull!(kvstore, idx, self.param_arrays[idx], priority=-idx)
+      else
+        # pull back the sum-ed gradients, to the same locations
+        pull!(kvstore, idx, self.grad_arrays[idx], priority=-idx)
+      end
+    end
+
+    if !update_on_kvstore
+      # manual updating
+      for i_dev = 1:num_dev
+        # create a fake index, so that the updater create states
+        # for different param AND different devices, TODO(mli)
+        # use a better solution later
+        fake_idx = idx * num_dev + i_dev
+        get(updater)(fake_idx, self.grad_arrays[idx][i_dev], self.param_arrays[idx][i_dev])
+      end
+    end
+  end
+end 
+
+""" 
+    get_params!(self, arg_params, aux_params)
+
+Copy data from each executor to `arg_params` and `aux_params`.
+# Arguments
+* `arg_params` : Dict{Symbol, Vector{NDArray}}. Target parameter arrays
+* `aux_params` : Dict{Symbol, Vector{NDArray}}. Target aux arrays
+
+# Notes
+This function will inplace update the NDArrays in arg_params and aux_params.
+"""
+function get_params!(self::DataParallelExecutorGroup, arg_params::Dict{Symbol, NDArray}, 
+                    aux_params::Dict{Symbol, NDArray})
+  for (name, block) in zip(self.param_names, self.param_arrays)
+    w = empty(size(block[1]))
```

Comment:
This needs to take type-inference into account. In effect there should be no un-typed `empty`, `zero`, or `one`